torch/_inductor/kernel/custom_op.py

```diff
@@ -0,0 +1,345 @@
+# Owner(s): ["module: inductor"]
+
+import functools
+from typing import Any, Callable, Optional, Union
+
+import torch
+from torch._inductor.codegen.subgraph import SubgraphTemplate
+from torch._inductor.ir import Buffer, FixedLayout, ir_node_to_tensor, TensorBox
+from torch._inductor.lowering import lowerings, validate_ir
+from torch._inductor.select_algorithm import (
+    autotune_select_algorithm,
+    ExternKernelChoice,
+)
+from torch._inductor.virtualized import V
+
+
+__all__ = [
+    "autotune_custom_op",
+    "register_custom_op_autotuning",
+]
+
+
+def _extract_tensor_inputs(
+    args: tuple[Any, ...], kwargs: dict[str, Any]
+) -> tuple[list[Any], dict[str, Any]]:
+    """Extract tensor inputs from mixed args/kwargs.
+
+    Args:
+        args: Positional arguments (mix of tensors and scalars)
+        kwargs: Keyword arguments (mix of tensors and scalars)
+
+    Returns:
+        Tuple of (tensor_inputs_list, non_tensor_kwargs)
+    """
+    tensor_inputs = []
+    non_tensor_kwargs = {}
+
+    for arg in args:
+        if isinstance(arg, (TensorBox, Buffer)) or (
+            hasattr(arg, "dtype") and hasattr(arg, "shape")
+        ):
+            tensor_inputs.append(arg)
+
+    for key, value in kwargs.items():
+        if isinstance(value, (TensorBox, Buffer)) or (
+            hasattr(value, "dtype") and hasattr(value, "shape")
```

Comment:
when are these `hasattr(value, "dtype") and hasattr(value, "shape")` necessary ? can we delete them?